docs: add detailed loadClass() TypeScript usage guide and new loadcla…

Necro-Rohan opened the pull request:
**PR Title:**
`Docs: Add guidance on using loadClass with TypeScript (fixes #12813)`

**PR Description:**

**Summary**

This PR addresses [[issue #12813](https://github.com/Automattic/mongoose/issues/12813)](https://github.com/Automattic/mongoose/issues/12813), which pointed out that it's confusing for developers to figure out how to use `schema.loadClass()` correctly with TypeScript.

To fix this, I've added a new documentation section to **`docs/typescript/statics-and-methods.md`** that provides clear, official guidance.

This new section covers the complete pattern discussed in the issue thread, including:
* **The Basic Pattern:** How to manually combine types for the Schema, Class, Model (`Model<Combined> & typeof Class`), and Document (`Document & Combined`).
* **Typing `this`:** The correct way to type the `this` parameter in both instance and static methods.
* **Known Limitations:** Explicitly documents the main caveats:
    * That `this` parameters aren't supported in getters/setters (ts(2784)).
    * The "lean" caveat: how `toObject()`, `toJSON()`, or `.lean()` return plain objects without the class methods, which can lead to runtime errors.

To validate this new documentation and prevent future regressions, this PR also adds a new `tsd` test file at **`test/types/docs-loadClass.test.ts`**.

**Examples**

The new documentation page itself contains all the code examples:
* See the new `## Using schema.loadClass()` section in **`docs/typescript/statics-and-methods.md`**.

The new `tsd` test file also serves as a practical, working example of these patterns:
* See **`test/types/docs-loadClass.test.ts`**.

---
✅ Adds documentation
✅ Adds type tests
✅ Fixes #12813

hasezoey commented on docs/typescript/statics-and-methods.md:
```diff
@@ -80,3 +80,186 @@ const doc = new User({ name: 'test' });
+
+
+
+# Using `loadClass()` with TypeScript
```
```suggestion

# Using `loadClass()` with TypeScript
```

Only one is enough, otherwise please put it into another file.

Maybe it would be a good idea to put this whole section into <https://mongoosejs.com/docs/advanced_schemas.html> instead.

hasezoey commented on docs/typescript/statics-and-methods.md:
```diff
@@ -80,3 +80,186 @@ const doc = new User({ name: 'test' });
+
+# Using `loadClass()` with TypeScript
+
+Mongoose supports applying ES6 classes to a schema using `schema.loadClass()`.
```
A link to the `loadClass` function would likely be good here.
<https://mongoosejs.com/docs/api/schema.html#Schema.prototype.loadClass()>

hasezoey commented on docs/typescript/statics-and-methods.md:
````diff
@@ -80,3 +80,186 @@ const doc = new User({ name: 'test' });
+
+```ts
+import { Schema, Model, model, Document } from 'mongoose';
+```
````
I dont think this is necessary. Also in its current state, this is a somewhat undocumented codeblock.

hasezoey commented on docs/typescript/statics-and-methods.md:
```diff
@@ -80,3 +80,186 @@ const doc = new User({ name: 'test' });
+
+---
+
+## Recommended Pattern
```
```suggestion
## Full example Code
```

I dont think `Recommended Pattern` quite fits this section, though i dont know if my suggestion is much better.

Necro-Rohan replied:
Sounds good. I've renamed that section heading to `## Full Example`.

hasezoey commented on test/types/loadclass.test.ts:
```diff
@@ -0,0 +1,178 @@
+  }
+}
+
+export {};
```
```suggestion
```

This shouldnt be necessary in the test file.

Necro-Rohan replied:
OK, I've removed the empty `export {}`.

Necro-Rohan commented on docs/typescript/statics-and-methods.md:
````diff
@@ -80,3 +80,186 @@ const doc = new User({ name: 'test' });
+);
+```
+
+```ts
````
Good suggestion, thanks! I've combined those code blocks.

Necro-Rohan commented on docs/typescript/statics-and-methods.md:
````diff
@@ -80,3 +80,186 @@ const doc = new User({ name: 'test' });
+doc.property1;         // schema prop
+```
+
+---
````
You're right, the headings are sufficient. I've removed the commented horizontal lines to clean up the formatting. Thanks!

Necro-Rohan commented on test/types/loadclass.test.ts:
```diff
@@ -0,0 +1,178 @@
+  type MyCombinedDocument = Document & MyCombined;
+
+  // Cast schema to satisfy TypeScript
+  const MyModel = model<MyCombinedDocument, MyCombinedModel>('MyClass', schema as any);
```
Sure, I've copied those explanatory comments from the test file into the documentation's code block.

Necro-Rohan commented on test/types/loadclass.test.ts:
```diff
@@ -0,0 +1,178 @@
+    // So we show an example error here.
+    get myVirtual() {
+      // @ts-expect-error: getter does not support `this` typing
+      return this.property1;
```
Thanks, I've added a note to the documentation explaining the workaround for the getter limitation `(using const self = this as ...;)`.

Necro-Rohan commented on test/types/loadclass.test.ts:
```diff
@@ -0,0 +1,178 @@
+// Test that `toObject()` / `toJSON()` lose class behavior.
+// But TypeScript does NOT warn you about this.
+//
+// This matches the behavior described in the issue:
```
Done, I've updated the comment to reference the issue number.

Necro-Rohan commented:
Thanks again for all the detailed feedback @hasezoey! I've just pushed an update that addresses all the suggestions and changes you asked for.

I've also fixed those `markdownlint` errors, so the CI should be passing now.

Please let me know what you think when you get a chance! I'm happy to make any other adjustments if needed.

hasezoey commented on docs/typescript/statics-and-methods.md:
```diff
@@ -80,3 +80,196 @@ const doc = new User({ name: 'test' });
+
+# Using `loadClass()` with TypeScript
+
+Mongoose supports applying ES6 classes to a schema using [`schema.loadClass()`](https://mongoosejs.com/docs/api/schema.html#Schema.prototype.loadClass()).
```
```suggestion
Mongoose supports applying ES6 classes to a schema using [`schema.loadClass()`](../api/schema.html#Schema.prototype.loadClass()).
```

Necro-Rohan commented on test/types/loadclass.test.ts:
```diff
@@ -0,0 +1,176 @@
+    // So we show an example error here.
+    get myVirtual() {
+      // @ts-expect-error: getter does not support `this` typing
+      return this.property1;
```
You're right, that should be in there. My mistake for overlooking it and the `Horizontal Rule's (<hr>)`. i have fixed it now.

vkarpov15 commented on docs/typescript/statics-and-methods.md:
```diff
@@ -80,3 +80,186 @@ const doc = new User({ name: 'test' });
+type MyCombinedModel = Model<MyCombined> & typeof MyClass;
+
+// A document must combine Mongoose Document + class + schema
+type MyCombinedDocument = Document & MyCombined;
```
Use HydratedDocument<> instead here

vkarpov15 commented on docs/typescript/statics-and-methods.md:
````diff
@@ -80,3 +80,186 @@ const doc = new User({ name: 'test' });
+To get full type support, you must manually merge types.
+
+```ts
+interface MySchema {
````
Prefer to call this `RawDocType` or something similar since this isn't a mongoose Schema

vkarpov15 reviewed (commented):
Overall I think something more like the following would be better, @hasezoey WDYT?

```ts
interface RawDocType {
  property1: string;
}

// `loadClass()` does NOT update TS types automatically.
// So we must manually combine schema fields + class members.
// The model type must include statics from the class
type MyCombinedModel = Model<RawDocType, {}, Pick<MyClass, 'myMethod'>, Pick<MyClass, 'myVirtual'>> & Pick<typeof MyClass, 'myStatic'>;

// A document must combine Mongoose Document + class + schema
type MyCombinedDocument = mongoose.HydratedDocument<RawDocType, Pick<MyClass, 'myMethod'>, {}, Pick<MyClass, 'myVirtual'>>;

const schema = new Schema({ property1: String });
schema.loadClass(MyClass);

// Cast schema to satisfy TypeScript
const MyModel = model<MyCombinedDocument, MyCombinedModel>(
  'MyClass',
  schema 
);
```

Another alternative I'd like to consider is to is recommending using `extends` on model instances as a way of handling methods, statics, and virtuals. The following works correctly for methods and statics, but has some caveats with virtuals.

```ts
class MyClass extends TestModel {
  myMethod() {
    return 'hello from myMethod ' + this.name;
  }

  static myStatic() {
    return 'hello from myStatic';
  }

  get myVirtual() {
    return 'hello from myVirtual';
  }
}

const doc = new MyClass({ name: 'test' });
// All work
console.log(doc.myMethod());
console.log(MyClass.myStatic());
console.log(doc.myVirtual);

// Caveat: output does NOT include myVirtual because myVirtual technically isn't a virtual here
console.log(doc.toObject());
```

hasezoey reviewed (approved):
Looks good to me in its current state as a initial guide. I agree with @vkarpov15's suggestions though.

FYI:
> `type MyCombinedModel = Model<RawDocType, {}, Pick<MyClass, 'myMethod'>, Pick<MyClass, 'myVirtual'>> & Pick<typeof MyClass, 'myStatic'>;`

Those `Pick`s are what i meant with "would require a Filter type", or a lot of manual typing.

Necro-Rohan commented on docs/typescript/statics-and-methods.md:
```diff
@@ -185,31 +191,20 @@ get myVirtual() {
-However, `toObject()` and `toJSON()` return **plain JavaScript objects** without these methods.
+When using the correct [Model](../api/model.html) and [HydratedDocument](../typescript.html) generics (as shown above), Mongoose's types for `toObject()` and `toJSON()` work as expected. They will correctly return a type that includes **only the raw data** (e.g., `RawDocType`), not the methods or virtuals.
+
+This is a feature, as it accurately reflects the plain object returned at runtime and prevents you from trying to access methods that don't exist.
```
Now that you’ve pointed it out, it does read a bit awkward. I’ll remove that part.

Necro-Rohan commented on docs/typescript/statics-and-methods.md:
```diff
@@ -225,41 +220,63 @@ TS cannot detect when class methods are dropped.
 
-If you want better type inference, [`methods`](../guide.html#methods) & [`statics`](../guide.html#statics) on schema are recommended.
+* ✅ works fine
+* ⚠ requires manual TS merging
```
done

Necro-Rohan commented on docs/typescript/statics-and-methods.md:
```diff
@@ -80,3 +80,202 @@ const doc = new User({ name: 'test' });
+| Automatic TS merging                           | ❌         |
+| `this` typing in methods                       | ✅         |
+| `this` typing in getters/setters               | ❌         |
+| Methods preserved in `toObject()` / `toJSON()` | ❌         |
```
Excellent point, The `toObject() and lean()` behaviors aren’t actually “limitations” of `loadClass()` they’re just how Mongoose works. I’ve removed those rows to avoid any confusion, and I’m also renaming the table to “**Feature Support Summary**” since “Limitations” felt a bit misleading.

Necro-Rohan commented:
@vkarpov15 I’ve made the updates based on your feedback.

- Removed the redundant toObject() / toJSON() section
- Renamed Limitations to Feature Support Summary and removed the toObject / lean rows

The PR has been updated and is ready for another look whenever you have a moment. Thanks again!
